src/rvsdg/from_tree.rs

```diff
@@ -0,0 +1,412 @@
+    typechecker::TypeCache,
+};
+
+use super::{BasicExpr, Operand, RvsdgBody, RvsdgFunction, RvsdgProgram, RvsdgType};
+
+type Operands = Vec<Operand>;
+
+/// State needed to translate a tree program to
+/// an RVSDG.
+struct TreeToRvsdg<'a> {
+    program: &'a TreeProgram,
+    /// A cache of types for every expression in program
+    type_cache: &'a TypeCache,
+    nodes: &'a mut Vec<RvsdgBody>,
+    /// The current state edge, threaded through
+    /// as the last arg to all stateful operations.
+    current_state_edge: Operand,
+    /// The current arguments to the tree program
+    /// as RVSDG operands. (doesn't include state edge)
+    current_args: Vec<Operand>,
+}
+
+pub(crate) fn tree_to_rvsdg(tree: &TreeProgram) -> RvsdgProgram {
+    let mut res = RvsdgProgram { functions: vec![] };
+    for func in &tree.functions {
+        res.functions.push(tree_func_to_rvsdg(func.clone(), tree));
+    }
+    res.functions
+        .push(tree_func_to_rvsdg(tree.entry.clone(), tree));
+    res
+}
+
+fn bril_type_from_type(ty: Type) -> bril_rs::Type {
+    match ty {
+        Type::Base(base_ty) => match base_ty {
+            BaseType::IntT => bril_rs::Type::Int,
+            BaseType::BoolT => bril_rs::Type::Bool,
+        },
+        Type::PointerT(ty) => {
+            let base_ty = bril_type_from_type(Type::Base(ty));
+            bril_rs::Type::Pointer(Box::new(base_ty))
+        }
+        Type::TupleT(_) => panic!("Tuple types not supported in RVSDG"),
+        Type::Unknown => panic!("Unknown type in tree_type_to_rvsdg_types"),
+    }
+}
+
+fn convert_func_output_type(ty: Type) -> Option<bril_rs::Type> {
+    match ty {
+        Type::TupleT(inner) => {
+            assert_eq!(
+                inner.len(),
+                0,
+                "Expected no tuple types in function_type_from_type"
+            );
+            None
+        }
+        _ => Some(bril_type_from_type(ty)),
+    }
+}
+
+fn tree_func_to_rvsdg(func: RcExpr, program: &TreeProgram) -> RvsdgFunction {
+    let output_type = func.func_output_ty().expect("Expected function types");
+
+    let Type::TupleT(input_types) = func.func_input_ty().expect("Expected function types") else {
+        panic!("Expected tuple type for inputs in tree_func_to_rvsdg")
+    };
+
+    let mut nodes = vec![];
+    let (typechecked_program, type_cache) = program.with_arg_types_and_cache();
+
+    let mut converter = TreeToRvsdg {
+        program: &typechecked_program,
+        type_cache: &type_cache,
+        nodes: &mut nodes,
+        // initial state edge is last argument
+        current_state_edge: Operand::Arg(input_types.len()),
+        // initial arguments are the first n arguments
+        current_args: (0..input_types.len()).map(Operand::Arg).collect(),
+    };
+
+    let converted = converter.convert_expr(func.clone());
+
+    let resulting_state_edge = converter.current_state_edge;
+    drop(converter);
```

Comment:
question: why do we need the drop? Does NLL not handle this for you?

Reply:
We don't need it, I deleted it.
I was confused at some point.